### PR TITLE
Add AsyncNotifyWhenDone

### LINKED ISSUE #2742: Server side API for waiting for cancellation.
Something like WaitTillDone(void* tag) or NotifyWhenDone?

### PR DESCRIPTION
**Pull request by yang-g**
Fixes #2742

### CODE REVIEW

**Comment by yang-g** [comment]
@a11r This is the API we discussed for async server side cancellation monitoring.

File: include/grpc++/server_context.h [hunk 1]
```diff
@@ -125,6 +125,11 @@ class ServerContext {
 
   const struct census_context* census_context() const;
 
+  // Async only. Has to be called before the rpc starts.
+  // Returns the tag in completion queue when the rpc finishes.
+  // IsCancelled() can then be called to check whether the rpc was cancelled.
+  void AsyncNotifyWhenDone(void* tag) { async_notify_when_done_tag_ = tag; }
```
Comment (ctiller) [inline include/grpc++/server_context.h]: It's legal to have a NULL tag... we should probably also have a `bool have_tag_` field here.

Comment (yang-g) [inline include/grpc++/server_context.h]: Done.

**Comment by ctiller** [comment]
Sanity is failing (Jenkins, not mine).

**Comment by yang-g** [comment]
Re-pushed.

**Pull request merged**